### PR TITLE
URL whitelist service

### LINKED ISSUE #6853: Implement URL white list service
## Description

When non-admin users have permissions to configure a lookup data adapter, they can use the `HTTPJSONPath` adapter to query potentially sensitive information. (e.g. EC2 instance metadata in AWS) The `HTTPJSONPath` data adapter executes a HTTP request for an URL that is defined by the user that configures the data adapter.

This is **only** an issue if an admin user is giving lookup-table permissions to a regular user. This is not the default.

To mitigate this issue, we want to implement an URL whitelist service. This service can be used by other subsystems that deal with user defined URLs to check if an URL is allowed to be used. The whitelist service requires a separate set of permissions to make sure users can be given permission to create/modify lookup tables but are not allowed to modify the URL whitelist.

## Tasks

- [x] Must be enabled by default
- [x] Support different match types
  - Regexp
  - Literal string
- [x] Separate permissions for the whitelist service
- [x] Check URLs at runtime
- [x] Configuration UI on "System/Configuration" page
- ~~[UI form component for URL whitelist that shows good error message and option to add new URL to whitelist (if user has permission).~~ This was moved to #7169
- [x] Use white list service in lookup data adapters (e.g. `HTTPJSONPath`, `DSVHTTP`) and event notifications
- [x] Create system notification when a runtime URL check failed
- [x] Create migration to add URL from known entities to the white list (e.g. lookup adapters, event notifications)
- [x] Backport to 3.1.x
- [x] Content pack support
- [x] Documentation

### PR DESCRIPTION
**Pull request by thll**
## Description

There are data adapters and event notifications which access URLs. These URLs will now have to be whitelisted in order for the requests to succeed.
This pull request adds a whitelist service, a corresponding REST API and a migration to provide this functionality.

## Motivation and Context

Fixes #6853

## How Has This Been Tested?

Whitelist created and updated via REST API. Migration executed.

Further extensive testing required once we are happy with the approach.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/lookup/adapters/DSVHTTPDataAdapter.java [hunk 1]
```diff
@@ -297,4 +320,17 @@ public boolean isCheckPresenceOnly() {
             public abstract DSVHTTPDataAdapter.Config build();
         }
     }
+
+    private void publishSystemNotificationForWhitelistFailure() {
```
[the remark this answers is not in the document]
Comment (thll) [reply graylog2-server/src/main/java/org/graylog2/lookup/adapters/DSVHTTPDataAdapter.java]: yep, makes sense.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/UrlWhitelistResource.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    @Inject
+    public UrlWhitelistResource(final UrlWhitelistService urlWhitelistService) {
+        this.urlWhitelistService = urlWhitelistService;
+    }
+
+    @GET
+    @Timed
+    @ApiOperation(value = "Get url whitelist.")
+    public UrlWhitelist get() {
+        checkPermission(RestPermissions.URL_WHITELIST_READ);
```
[the remark this answers is not in the document]
Comment (thll) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/system/UrlWhitelistResource.java]: That's handy. Will do that.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/system/UrlWhitelistResource.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @POST
+    @Path("/check")
+    @Timed
+    @ApiOperation(value = "Check if a url is whitelisted.")
+    @NoAuditEvent("Validation only")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public WhitelistCheckResponse check(@ApiParam(name = "url", required = true)
+                             @Valid /*@NotNull*/ final WhitelistCheckRequest checkRequest) {
```
[the remark this answers is not in the document]
Comment (thll) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/system/UrlWhitelistResource.java]: Nope, shouldn't be commented. The null check is actually needed.

File: graylog2-server/src/main/java/org/graylog2/migrations/V20191129134600_CreateInitialUrlWhitelist.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        return whitelist;
+    }
+
+    private Optional<WhitelistEntry> extractFromNotification(NotificationDto notificationDto) {
+        final EventNotificationConfig config = notificationDto.config();
+
+        if (!(config instanceof HTTPEventNotificationConfig)) {
+            return Optional.empty();
+        }
+        final String url = ((HTTPEventNotificationConfig) config).url();
```
[the remark this answers is not in the document]
Comment (thll) [reply graylog2-server/src/main/java/org/graylog2/migrations/V20191129134600_CreateInitialUrlWhitelist.java]: I've added that and also that the whitelist is actually checked when the callback is configured or executed. I didn't add a system notification here though, as it didn't turn out to be straight forward. Can we do without it? After all the code is considered _legacy_.

**Comment by thll** [comment]
I've implemented the requested changes and also sneaked in a little refactoring. Please have a look.

File: graylog2-server/src/test/java/org/graylog2/system/urlwhitelist/UrlWhitelistServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import org.mockito.MockitoAnnotations;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableList;
+
+import java.util.Collections;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class UrlWhitelistServiceTest {
+    @Mock
+    ClusterConfigService clusterConfigService;
```
Comment (danotorrey) [inline graylog2-server/src/test/java/org/graylog2/system/urlwhitelist/UrlWhitelistServiceTest.java]: Can this field be removed?

File: graylog2-server/src/main/java/org/graylog2/contentpacks/facades/UrlWhitelistFacade.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+import javax.inject.Inject;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.graylog2.contentpacks.model.ModelTypes.URL_WHITELIST_ENTRY_V1;
+
+public class UrlWhitelistFacade implements EntityFacade<WhitelistEntry> {
+    private static final Logger LOG = LoggerFactory.getLogger(UrlWhitelistFacade.class);
```
Comment (danotorrey) [inline graylog2-server/src/main/java/org/graylog2/contentpacks/facades/UrlWhitelistFacade.java]: Logger is not used. Not a big deal through.

File: graylog2-server/src/main/java/org/graylog2/lookup/adapters/DSVHTTPDataAdapter.java [hunk 7]
```diff
@@ -104,16 +113,27 @@ public Duration refreshInterval() {
 
     @Override
     protected void doRefresh(LookupCachePurge cachePurge) throws Exception {
+        if (!whitelistService.isWhitelisted(config.url())) {
+            setError(UrlNotWhitelistedException.forUrl(config.url()));
+            publishSystemNotificationForWhitelistFailure();
+            return;
+        }
+
+        boolean urlWasNotWhitelisted = getError().filter(UrlNotWhitelistedException.class::isInstance).isPresent();
```
Comment (danotorrey) [inline graylog2-server/src/main/java/org/graylog2/lookup/adapters/DSVHTTPDataAdapter.java]:
```suggestion
        final boolean urlWasNotWhitelisted = getError().filter(UrlNotWhitelistedException.class::isInstance).isPresent();
```

File: graylog2-server/src/test/java/org/graylog2/system/urlwhitelist/UrlWhitelistServiceTest.java (hunk 5 above)
Comment (thll) [reply graylog2-server/src/test/java/org/graylog2/system/urlwhitelist/UrlWhitelistServiceTest.java]: yep

File: graylog2-server/src/main/java/org/graylog2/contentpacks/facades/UrlWhitelistFacade.java (hunk 6 above)
Comment (thll) [reply graylog2-server/src/main/java/org/graylog2/contentpacks/facades/UrlWhitelistFacade.java]: Right. Removed it.

File: graylog2-server/src/main/java/org/graylog2/lookup/adapters/DSVHTTPDataAdapter.java (hunk 7 above)
Comment (thll) [reply graylog2-server/src/main/java/org/graylog2/lookup/adapters/DSVHTTPDataAdapter.java]: Changed that.

**Pull request merged**